### DISCUSSION TITLE
Out-of-process caches

### DISCUSSION THREAD

**Discussion by michaelhixson** [TechEmpower/FrameworkBenchmarks#5844] [RFCs]
For the recently-added [cached query](https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Project-Information-Framework-Tests-Overview#caching) test type, we expect two main categories of caches:

* In-process / in-memory caches.
* Out-of-process / distributed caches.

This issue is about out-of-process caches.

Those out-of-process caches can run in three ways:

* Same container.
* Separate container, same host.
* Separate host.

Currently, only the first one is achievable without modifying the toolset, so that's what some frameworks have implemented.  We feel this is the least realistic of the three ways to run an out-of-process cache.

Therefore, we propose the following:

* Modify the toolset to enable separate container caches.
* Support specific cache servers the same way we support specific databases.
* Support Redis and Memcached at first.
* The cache runs on either the application host or the database host.  The test implementor chooses which.
* A benchmark_config.json entry indicating a cached query implementation might include lines like this:
  ```json
  "cached_query_url": "/cached_query?queries=",
  "cache_server": "Memcached",
  "cache_server_location": "Local",
  ```
* Disallow same container, out-of-process caches.  Migrate existing implementations to use separate container caches.

Comments, questions, and alternative suggestions are welcome.

**Comment by jcheron** [comment]
I have a question,
just to make sure to correctly interpret the specifications of the **Caching** test:

> iv. The request handler must retrieve a set of CachedWorld objects, equal in count to the count parameter, from the CachedWorld database table. Note that the CachedWorld database table uses a schema idential to the World table used for Test #3 but is distinct to avoid confusion.

I interpret it this way: 
- For each query, the random loading of the count **cachedWorld** objects is done as follows:
  - For each object:
    If the instance is present in cache, recovery of this instance, otherwise, loading from the database and caching.

> x. As with other test types, a warm-up will occur prior to capturing performance data. However, it is also acceptable for an implementations to prime its cache during setup in whatever form needed, assuming that priming process completes in a reasonable amount of time.

The cache can be completely initialized at the startup of the application, by loading the objects from the database.

Some implementations of the Caching test perform this caching preload, but do not interpret point **iv** in the same way as I do, by fetching the cached objects directly at each request.
In other words, if we externalize the cache, we load the 10000 instances of CacheWorld on a Redis server, and then, at each request, the http server queries the Redis server to randomly retrieve the data (without possibly querying the database server).
This is no longer testing a database cache, but querying a cache system.

Other implementations go even further, and do without any caching system, by storing all instances of CachedWorld in a variable, and then using this variable for each request. This method avoids the unserialization step.

What is the correct interpretation?

**Comment by jcheron** [comment]
I can ask my question in a new issue if you wish, since I'm going a little off topic.

**Comment by michaelhixson** [comment]
> What is the correct interpretation?

I think you interpreted it correctly, unless I'm missing something subtle.  Are you hinting that you'd like to prohibit cache priming and ensure some number of cache misses during the test?

**Comment by msmith-techempower** [comment]
> The cache can be completely initialized at the startup of the application, by loading the objects from the database.

Correct

> the http server queries the Redis server to randomly retrieve the data (without possibly querying the database server).

Correct

> Other implementations go even further, and do without any caching system, by storing all instances of CachedWorld in a variable, and then using this variable for each request.

Correct

> What is the correct interpretation?

EDIT: (I never answered the question) I do not believe there is a "correct" interpretation. Each solution has a valid use case in real-world applications and often one will find a mix, or all, of these solutions is appropriate for a given project. Perhaps, a better answer would be that they are *all* "correct", and we need to distinguish the solutions for consumption (read: results website).

We are working on fleshing out some flags to distinguish each of these scenarios. At present, we plan on supporting them all as `cached_query` test implementations, but want to distinguish each "flavor". If it ends up being, and I expect it will, that caching the entire table in memory on the application is more performant than making a network request to a Redis instance, for example, then we should be forthcoming about the difference in implementation.

At present, and this may change, we envision describing the flavors as: 1. in memory, 2. cache on app server, 3. cache on database server.

**Comment by jcheron** [comment]
Thank you for your answers.
I am not opposed to any solution, as long as each of them provides answers to a common use case in production.
I was mainly looking for the right way to implement, respecting the specifications.

But the direct query of a cache system (which assumes static data) is different and does not correspond to the same needs as the implementation of a database cache system, as well as the use of an in-memory cache, which requires a measured volume of data.
It is a good idea to flag these different approaches.

**Comment by matt-42** [comment]
Is it ok to cache all world objects in an array so that array[i] gives direct access to the world object with id i ?
It's written in the requirement "You cannot use plain key/value map" but you do not forbid plain arrays, which are even more efficient.

**Comment by msmith-techempower** [comment]
>  It's written in the requirement "You cannot use plain key/value map" but you do not forbid plain arrays, which are even more efficient.

This is intended to be disallowed for the same reason.

**Comment by matt-42** [comment]
I was asking because some tests are using them:
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Go/atreugo/src/views/views.go#L23
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Go/fasthttp/src/handlers/handlers.go#L25
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Go/fiber/src/server.go#L164

**Comment by msmith-techempower** [comment]
> I was asking because some tests are using them:

I don't know Go very well, but I agree that those appear to be using plain old arrays, which I *believe* we want to expressly disallow.

ping @michaelhixson for clarification on the rules.

**Comment by matt-42** [comment]
And I think we also need a definition of 'plain key/value map and array'. What features do you require for it to be non-plain ?

**Comment by dantti** [comment]
Ping, I'd like to implement this test also, but my db cache mechanism internaly uses a key (query + parameters) / value (result + timestamp) pair, I do use this on production as thanks to Postgres notifications I can invalidade the cache only when the data changes. Is this allowed?
